libtiledbsoma/src/soma/soma_array.cc

```diff
@@ -647,40 +631,126 @@ uint64_t SOMAArray::nnz() {
+        if (current_range[1] < non_empty_domains[i][0]) {
+            // Fragment is non-overlapping.
+
+            // Current range was overlapping with previous tiles
+            // and we need to add it to the overlapping ranges.
+            if (overlap) {
+                overlapping_ranges.push_back(
+                    std::make_pair(current_range[0], current_range[1]));
+            } else {
+                auto frag_ts = fragment_info.timestamp_range(
+                    relevant_fragments[i - 1]);
+
+                if ((timestamp_ && !(frag_ts.first >= timestamp_->first &&
+                                     frag_ts.second <= timestamp_->second)) ||
+                    (!schema_->allows_dups() &&
+                     frag_ts.first != frag_ts.second)) {
+                    // fragment overlaps read timestamp range, but isn't fully
+                    // contained within: fall back to count_cells to sort that
+                    //                          or
+                    // If any relevant fragment is a consolidated fragment, fall
+                    // back to counting cells, because the fragment may contain
+                    // duplicates. If the application is allowing duplicates (in
+                    // which case it's the application's job to otherwise ensure
+                    // uniqueness), then sum-over-fragments is the right thing
+                    // to do.
+
+                    overlapping_ranges.push_back(
+                        std::make_pair(current_range[0], current_range[1]));
+                } else {
+                    total_cell_num += cell_count;
+                }
+            }
+
+            current_range = non_empty_domains[i];
+            cell_count = fragment_info.cell_num(relevant_fragments[i]);
+            overlap = false;
+        } else if (current_range[1] >= non_empty_domains[i][1]) {
+            // Check if range is included completely
+            overlap = true;
+        } else if (current_range[1] < non_empty_domains[i][1]) {
+            // Check if range is partially overlapping
+            current_range[1] = non_empty_domains[i][1];
             overlap = true;
-            break;
         }
     }
 
-    // If relevant fragments do not overlap, return the total cell_num
-    if (!overlap) {
-        return total_cell_num;
+    if (overlap) {
+        overlapping_ranges.push_back(
+            std::make_pair(current_range[0], current_range[1]));
+    } else {
+        auto frag_ts = fragment_info.timestamp_range(relevant_fragments.back());
+
+        if ((timestamp_ && !(frag_ts.first >= timestamp_->first &&
+                             frag_ts.second <= timestamp_->second)) ||
+            (!schema_->allows_dups() && frag_ts.first != frag_ts.second)) {
+            overlapping_ranges.push_back(
+                std::make_pair(current_range[0], current_range[1]));
+        } else {
+            total_cell_num += cell_count;
+        }
```

Comment:
This looks similar to L650-L676. Can this be pulled out into a separate function?